### DISCUSSION TITLE
`@` and `VecSS`

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#1954] [Ideas & RFCs]
```rs
VecSS = Vec<StringOrSection>  // header value, or the condition portion of conditional header
```

```
-- foo: {
    this is a long
    caption that spans multiple lines
}
next-foo-header: 1
```

```
-- foo: 
color if { 
    x > 10 
}: red
```

```
-- foo: 
color if { 
    x > {
    10  + 20 }
} > { 20 }: red
```

```
-- foo: hello {

    -- bar: 

} world {
    -- hello: @x: 20
    -- hello: \@x: 20
}
next-foo-header: 1
```

```
-- foo: hello {-- bar:} world {-- hello:}
next-foo-header: 1
```

**Comment by amitu** [comment]
Instead of 

```
-- foo: x
if: $var
```

we can do

```
-- if $var:

   -- foo: x
   -- bar:

-- end: if
```

Which will go with:

```
-- match $x:

-- case 10: 

    -- show-10:
    -- foo:

-- end: match
```

Also, instead of

```
-- foo: $x
for: x in $list
```

We can do:

```
-- for (x, index): $list

-- foo: $x
-- bar: $index

-- end: for
```

**Reply by siddhantk232** [reply]
How do you translate it into our grammar? Do you even take that into consideration?

```
-- for (x, index): $list

-- foo: $x
-- bar: $index

-- end: for
```

According to the grammar you defined in the spec, `for` is the `section-kind` which is fine but `(x, index)` is the section name? I am not sure how I feel about that. This feels that the parser will left some details for later for us to parse differently (`(x, index)`). Same goes for `$list`, I think that is a `caption` and a caption is always a string? Which also means that our grammar will not be able to parse it?

**Reply by amitu** [reply]
Caption is not always a string, you can use any type as caption.

We will have to update our `section-kind section-name` syntax to something more fitting this.

**Reply by amitu** [reply]
Instead of 

```ftd
-- for (x, index): $list

-- foo: $x
-- bar: $index

-- end: for
```

Let's follow [Django's example](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#for):
<img width="766" alt="image" src="https://github.com/user-attachments/assets/a5e929b0-3a24-4d00-b7b7-ec88c1c38c14">

and create

```ftd
-- for x: $list

-- foo: $x
-- bar: $FORLOOP.index

-- end: for
```

**Comment by amitu** [comment]
What if we used `[]` for multiline string, both in `p-script` (headers/cation) and `f-script`.

```
-- foo: [
    this is a long
    caption that spans multiple lines
]
next-foo-header: 1
```

```
-- foo: 
color if { 
    x > 10 
}: red
```


```
-- foo: 
color if { 
    x > {
    10  + 20 }
} > { 20 }: red
```



```
-- foo: [hello {

    -- bar: 

} world {
    -- hello: @x: 20
    -- hello: \@x: 20
}]
next-foo-header: 1
```

```
-- foo: [hello {-- bar:} world {-- hello:}]  ;; in this the `[]` are optional
next-foo-header: 1
```

```
-- foo: [
      hello { x } ;; text inside inside curly does not start with `--`, so it is an expression, and 
                  ;; refers to a variable x. 
      { [x] } ;; [x] refers to a string literal "x" in f-script
      {-- bar:} 
      world {-- hello:} 
     { some-function() }
]  
next-foo-header: 1
```

**Reply by amitu** [reply]
Rule for `{}` becomes that `p-script` curly will always contains section or expression, if used inside string, and for non string types it will always contain expression. In the original description, for string, the curly contained raw string as well along with section and expression.
